src/NuGet.Core/NuGet.DependencyResolver.Core/Remote/RemoteWalkContext.cs

```diff
@@ -73,7 +73,21 @@ public IList<IRemoteDependencyProvider> FilterDependencyProvidersForLibrary(Libr
                     return Array.Empty<IRemoteDependencyProvider>();
                 }
 
-                return RemoteLibraryProviders.Where(p => sources.Contains(p.Source.Name)).AsList();
+                // PERF: Avoid Linq in hot paths.
+                var filteredLibraryProviders = new List<IRemoteDependencyProvider>();
+                for (int i = 0; i < RemoteLibraryProviders.Count; ++i)
+                {
+                    var current = RemoteLibraryProviders[i];
+                    for (int j = 0; j < sources.Count; ++j)
```

Comment:
We can optimistically allocate an array of sources length.

Not finding all sources implies that someone has a package source mapping configuration that points to a source that's disabled. 
While possible, not very likely.

Reply:
Wanted to double check, you mean that, in general, `filteredLibraryProviders` should be the same length as `sources` (i.e. `var filteredLibraryProviders = new List<IRemoteDependencyProvider>(sources.Count)`)?

Reply:
Yes. 

RemoteLibraryProviders represents all the sources, while sources represents the sources configured according to package source mapping.

Reply:
Updated :)